compact_str/src/repr/boxed/mod.rs

```diff
@@ -151,6 +151,20 @@ impl BoxString {
         }
     }
 
+    #[inline]
+    pub fn into_string(self) -> String {
+        // Ensure that we don't free the data we are transferring to `String`
+        let this = ManuallyDrop::new(self);
+
+        // SAFETY:
+        // * The memory in `ptr` was previously allocated by the same allocator the standard library
+        //   uses, with a required alignment of exactly 1.
+        // * `length` is less than or equal to capacity, due to internal invaraints.
+        // * `capacity` is correctly maintained internally.
+        // * `BoxString` only ever contains valid UTF-8.
+        unsafe { String::from_raw_parts(this.ptr.as_ptr(), this.len, this.capacity()) }
```

Comment:
A `BoxedString` can store the capacity on either the stack _or_ the heap, we store the capacity on the heap when there isn't enough room on the stack, see [this doc](https://github.com/ParkMyCar/compact_str/blob/main/compact_str/src/repr/boxed/capacity.rs#L18-L36) comment for an explanation.

tl;dr I think you'll have to change this to something like:
```
match self.cap.as_usize() {
  // capacity is on the stack, form a string from these parts
  Ok(capacity) => {
    let this = ManuallyDrop::new(self);
    unsafe { String::from_raw_parts(this.ptr.as_ptr(), this.len, capacity) }
  },
  // capacity is on the heap! we need to form a new String
  Err(_) => {
    // ... drop the existing string and form a new one
  }
}
```

Reply:
Good idea to re-use the existing buffer. I think this is something we can iterate on, for now I filed issue #120 to track this

Reply:
> the layouts aren't compatible

Yeah, you'll run into an issue as `String`'s buffer is `u8` aligned but the inline length is `usize` aligned. *At a performance cost*, we could store the length unaligned to make the allocated alignment 1.

The unstable `Allocator` trait allows *requesting* (though an implementation can theoretically refuse to fulfill) a change in alignment of an allocation. The stable `GlobalAlloc` does not.

My vote is to just copy/realloc, and put a note tl if/when `Allocator` is stabilized optimize it to opportunistically use a layout-adjusting realloc so the underlying allocator can avoid the copy.